Adjust chain-following code to account for collateral inputs

sevanspowell opened the pull request:
The purpose of this PR is to modify the the code that follows the blockchain to adjust the Wallet's UTxO when collateral inputs are lost due to a transaction that fails phase-2 script validation (Plutus scripts).

I have:
- Add valid script field to `MkApiTransactionParams`, `ApiTransaction`, `TxMeta`, `Tx`, and `TransactionInfo`.
- Modify state transition logic (when a Tx is applied to a UTxO) to account for collateral inputs.
- Add integration test for collateral inputs.
- Add property tests for state transition logic.
- Add is_valid_script field to API specification.

### Issue Number

ADP-1036

sevanspowell committed:
Add the "isValidScript" field to track script validation

- Add new field to track script validation in Alonzo-era transactions.
- Modify ApiTransaction type to return this new field and update the API
  accordingly.

sevanspowell commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2880,6 +2885,7 @@ data MkApiTransactionParams = MkApiTransactionParams
     , txMeta :: W.TxMeta
     , txMetadata :: Maybe W.TxMetadata
     , txTime :: UTCTime
+    , txIsValidScript :: Maybe Bool
```
I need to choose a more descriptive type for this:

Nothing -> Not an era that cares about validation
Just False -> Cares about validation, but validation failed
Just True -> Cares about validation and validation succeeded

sevanspowell replied:
06493c0

sevanspowell committed:
Update state transition rules to account for invalid scripts

- Transactions that make use of smart contracts and other scripts may fail to
  validate. Modify the state transition rules to correctly remove collateral
  inputs from the UTxO instead of regular inputs when script validation fails.
- Add a tests to verify this logic.

sevanspowell committed:
Use more descriptive type for isValidScript

- Use a more descriptive type for isValidScript, to prevent confusion.

sevanspowell commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -503,7 +512,9 @@ prefilterBlock b u0 = runState $ do
 
         return $ if hasKnownOutput && not hasKnownInput then
```
A Tx that fails will have no known output nor any known input, thanks to the implementation of `utxoFromTx`. This means it will hit the last `else` in this `if-else`, which is to just use the `prevUTxO`, but we also want to remove collateral from the UTxO...

I need to write a failing test to demonstrate what I mean.

sevanspowell replied:
I think this will suffice as an implementation:

```
        let hasKnownInput =
                (if failedScriptValidation (tx ^. #isValidScript)
                 then Set.fromList (collateralInputs tx)
                 else Set.fromList (inputs tx)
                ) `Set.intersection` (Set.fromList ownedAndKnownTxIns)
                /= mempty
```

sevanspowell replied:
Actually not quite sufficient. We can’t presume that we already know about our inputs that could be used for collateral. The transaction may introduce collateral inputs that we own, so we need to expand the right hand side of the Set intersection too.

sevanspowell replied:
It is surprising this logic is not tested at a unit level, it’s very particular.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -459,10 +478,31 @@ data TransactionInfo = TransactionInfo
+data ScriptValidation = ScriptsNotSupported
+                      | ScriptValidationPassed
+                      | ScriptValidationFailed
+  deriving (Generic, Show, Eq, Ord)
```
Suggestions:
1. Use the `Tx` prefix, since this relates to a transaction and appears in the `Tx` module.
2. Use `Validity`, as we can then say "what is the validity of the script for this transaction"?

As for the second point, we could use "validation status".

```suggestion
-- | Indicates whether the script associated with a transaction has passed or
-- failed validation. Pre-Alonzo era, scripts were not supported.
data TxScriptValidity
    = TxScriptsUnsupported
    -- ^ Indicates that scripts were not supported in the era in which this
    -- transaction was created.
    | TxScriptValid
    -- ^ Indicates that the script passed validation.
    | TxScriptInvalid
    -- ^ Indicates that the script failed validation.
  deriving (Generic, Show, Eq, Ord)
```

sevanspowell replied:
a6a26a6

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Gen.hs:
```diff
@@ -125,12 +132,18 @@ txToTxWithoutId :: Tx -> TxWithoutId
+    (TxWithoutId a1 a2 a3 a4 a5 a6 a7)
+
+genScriptValidation :: Gen ScriptValidation
+genScriptValidation = genericArbitrary
```
Here you could also use `arbitraryBoundedEnum`, which will generate a uniform distribution across all possibilities.

jonathanknowles commented on lib/core/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -1284,7 +1285,14 @@ dummyTransactionLayer = TransactionLayer
+                    (outputsCovered cs)
+                    mempty
+                    Nothing
+                    ScriptsNotSupported
```
Suggestion: use record construction syntax here. It's easier to guess what `Nothing` refers to if there's a field name provided!

sevanspowell replied:
I'm glad you said that, because the order of inputs and collateralInputs was incorrect!

sevanspowell replied:
95b37cf

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -2097,6 +2105,7 @@ components:
         mint: *transactionMint
         status: *transactionStatus
         metadata: *transactionMetadata
+        is_valid_script: *transactionScriptValidity
```
Suggestion: use `script_validity` here:
```suggestion
        script_validity: *transactionScriptValidity
```

sevanspowell replied:
f21ba22

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2880,6 +2885,7 @@ data MkApiTransactionParams = MkApiTransactionParams
     , txMeta :: W.TxMeta
     , txMetadata :: Maybe W.TxMetadata
     , txTime :: UTCTime
+    , txIsValidScript :: W.ScriptValidation
```
```suggestion
    , txScriptValidity :: W.TxScriptValidity
```

sevanspowell replied:
a6a26a6

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -111,6 +111,7 @@ TxMeta
     txMetadata              W.TxMetadata Maybe  sql=data
     txMetaSlotExpires       SlotNo Maybe        sql=slot_expires
     txMetaFee               Word64 Maybe        sql=fee
+    txMetaIsValid           Bool Maybe          sql=is_valid
```
```suggestion
    txMetaScriptValidity    Bool Maybe          sql=script_validity
```

sevanspowell replied:
a6a26a6

sevanspowell committed:
Rename ScriptValidation to something clearer

sevanspowell committed:
Remove debugging ToExpr instance

sevanspowell committed:
Use record constructor for Tx to aid readability

sevanspowell committed:
Rename "isValidScript" to "scriptValidity"

sevanspowell committed:
Fixup rebase

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -1674,6 +1674,14 @@ x-maximumTokenBundleSize: &maximumTokenBundleSize
+    validated or not. Previous to Alonzo-era, this is null.
+  nullable: true
+  type: boolean
+  default: null
```
Hi @sevanspowell

I think it might be better if this was an _enumeration_, rather than a Boolean.

For example:
```yaml
  x-txScriptValidity: &txScriptValidity
    type: string
    enum:
      - unsupported
      - valid
      - invalid
```

I think this has the following advantages:

1. We can generate the `ToJSON` and `FromJSON` instances generically, rather than having to write them manually (which requires maintenance), as it's now just a 1:1 mapping from the API type to the primitive type.
2. We don't need to deal with the missing case, as it's represented explicitly.
3. The individual cases are more self-documenting.

We also use enumerations elsewhere in the API to represent this kind of multi-option value. So it would be consistent with what we've done before. For example:

```hs
data ApiEra
    = ApiByron
    | ApiShelley
    | ApiAllegra
    | ApiMary
    | ApiAlonzo
    deriving (Show, Eq, Generic, Enum, Ord, Bounded)
    deriving anyclass NFData

instance FromJSON ApiEra where
    parseJSON = genericParseJSON $ Aeson.defaultOptions
        { constructorTagModifier = drop 4 . camelTo2 '_' }
instance ToJSON ApiEra where
    toJSON = genericToJSON $ Aeson.defaultOptions
        { constructorTagModifier = drop 4 . camelTo2 '_' }  
```

Thoughts?

sevanspowell replied:
Yeah, I think that's great. Implementing now...

sevanspowell replied:
1531fc7

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1979,6 +1990,11 @@ txHistoryFromEntity ti tip metas ins cins outs ws =
+                    case isValid of
+                        Nothing -> W.TxScriptsUnsupported
+                        Just False -> W.TxScriptInvalid
+                        Just True -> W.TxScriptValid
```
Formatting: use 4-space indents:

```suggestion
            , W.txInfoScriptValidity =
                case isValid of
                    Nothing -> W.TxScriptsUnsupported
                    Just False -> W.TxScriptInvalid
                    Just True -> W.TxScriptValid
```

sevanspowell replied:
6c327fadfe2f1ed417ba2e05f3d200ad87ae00fa

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -111,6 +111,7 @@ TxMeta
     txMetadata              W.TxMetadata Maybe  sql=data
     txMetaSlotExpires       SlotNo Maybe        sql=slot_expires
     txMetaFee               Word64 Maybe        sql=fee
+    txMetaScriptValidity    Bool Maybe          sql=script_validity
```
In some ways it would be nice to use an enumeration type here that matches the primitive type, as this would make it easier to add additional constructors in future, should we need to.

But that would presumably require a DB migration to get right. (Because the column would be missing for existing transactions.)

So I think it's the right choice to use a `Bool Maybe` here.

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -798,6 +807,29 @@ instance PersistField DerivationPrefix where
+instance PersistField TxScriptValidity where
+    toPersistValue TxScriptsUnsupported = PersistNull
+    toPersistValue TxScriptValid = PersistBool True
+    toPersistValue TxScriptInvalid = PersistBool False
```
Suggestion: use `LambdaCase` here for brevity:

```suggestion
    toPersistValue = \case
        TxScriptsUnsupported -> PersistNull
        TxScriptValid -> PersistBool True
        TxScriptInvalid -> PersistBool False
```

sevanspowell replied:
383227a

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -798,6 +807,29 @@ instance PersistField DerivationPrefix where
+          , " expected null or boolean"
+          , " from database, but received: "
+          , T.pack (show x)
+          ]
```
Suggestion: use `LambdaCase` here for brevity.

Also, using `T.unwords` will insert spaces for you.

```suggestion
    fromPersistValue = \case
        PersistNull       -> Right TxScriptsUnsupported
        PersistBool True  -> Right TxScriptValid
        PersistBool False -> Right TxScriptInvalid
        x -> Left $ T.unwords
            [ "Failed to parse Haskell type `ScriptValidation`;"
            , "expected null or boolean"
            , "from database, but received:"
            , T.pack (show x)
            ]
```

sevanspowell replied:
383227a

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -798,6 +807,29 @@ instance PersistField DerivationPrefix where
+
+instance PersistFieldSql TxScriptValidity where
+    -- sqlType _ = sqlType (Proxy @ScriptValidation)
+    sqlType _ = SqlBool
```
I think you can write this as:
```suggestion
    sqlType _ = sqlType (Proxy @Bool)
```

But maybe it should be:
```suggestion
    sqlType _ = sqlType (Proxy @(Maybe Bool))
```

sevanspowell replied:
2f5cbb5

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -373,7 +375,12 @@ applyTxToUTxO tx !u = spendTx tx u <> utxoFromTx tx
+        if failedScriptValidation (tx ^. #scriptValidity)
+        then Set.fromList (collateralInputs tx)
+        else Set.fromList (inputs tx)
+    )
```
Suggestion: extract out an `inputsToExclude` definition.

Advantage: avoid the multi-line parentheses, and avoid repeating the `Set.fromList`.

```suggestion
spendTx tx !u =
    u `excluding` Set.fromList inputsToExclude
  where
    inputsToExclude =
        if failedScriptValidation (tx ^. #scriptValidity)
        then collateralInputs tx
        else inputs tx
```

sevanspowell replied:
de801657f95c2097e387a83565de213271260ab6

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -556,4 +567,11 @@ changeUTxO
+    where
+        -- Generate a UTxO from an transaction, assuming that it passes phase-2
+        -- script validation. Crucially, utxoFromTx will exclude failed
+        -- transactions, hence we define our own function.
```
This leaves me with a question:

For transactions in the pending set, what meaning can we ascribe to the `txScriptValidity` field? It seems that at this stage, we wouldn't have enough information to assign a meaningful value. (Unless I'm mistaken?)

The fact that we must ignore the `txScriptValidity` field in the `changeUTxO` function indicates to me that there might be a better way to represent the concept of "not-yet-known" or "unknown".

For example, we could give the `TxScriptValidity` type just two options (`TxScriptValid` and `TxScriptInvalid`), and then give `txScriptValidity` a type of `Maybe TxScriptValidity`. The `Nothing` option would represent the case where we don't yet know. But this also seems non-ideal.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1019,8 +1028,12 @@ fromAlonzoTx
+    (\(tx, d, p) -> (tx { W.scriptValidity = if isValid
+                                            then W.TxScriptValid
+                                            else W.TxScriptInvalid
+                        }, d, p))
```
Request: try to avoid using variable-length indentation. (See https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#avoid-variable-length-indentation)

Suggestion:
```suggestion
    (\(tx, d, p) -> (tx {W.scriptValidity = validity}, d, p))
    $ fromAlonzoTxBodyAndAux bod aux
  where
    validity =
        if isValid
        then W.TxScriptValid
        else W.TxScriptInvalid
```

sevanspowell replied:
3997bf9

sevanspowell committed:
Use generic ToJSON/FromJSON instances

jonathanknowles reviewed (commented):
@sevanspowell and I had a discussion about the representation of a transaction's script validity (currently represented by the `Tx.txScriptValidity :: TxScriptValidity` field).

After consideration, we both feel that it doesn't make sense to talk about the script validation status of a transaction that is still in the pending set: after all, transactions in this state have not yet been subjected to phase-2 validation, so we can't yet know what the outcome will be.

Therefore, we propose to:

- [x] Change the type of `Tx.txScriptValidity` to `Maybe TxScriptValidity`.
    (This makes it possible to assign pending transactions with a script validity of `Nothing`.)
- [x] Change the type of `TxInfo.txInfoScriptValidity` to `Maybe TxScriptValidity`.
- [x] Change the definition of `TxScriptValidity` to:
    ```hs
    data TxScriptValidity
        = TxScriptValid
        | TxScriptInvalid
    ```
- [x] Change the API definition of `txScriptValidity` to:
    ```yaml
    x-txScriptValidity: &txScriptValidity
      type: string
      enum:
        - valid
        - invalid
    ```
The absence of a validity means that either:
- script validity is not supported for this type of transaction (for example, if the transaction was created in an era prior to Alonzo); or
- this transaction has not yet been subjected to script validation.

We both feel that this representation is still not completely satisfactory. Even under the revised representation, the `Tx` type still make it possible to represent a pending transaction that has a script validity, which of course doesn't make sense.

Ideally, we'd want to represent pending transactions with a type that **_doesn't_** have a script validity field, so it would not be possible to represent this invalid state.

However, the `Tx` type is deeply embedded in multiple places, and replacing this with a `PendingTx` type would likely be a lot of work, and almost certainly cause us to exceed our time budget for this task. So we propose to:

- [x] Create an item of technical debt: represent pending transactions with a type that doesn't include script validity.

sevanspowell committed:
Change representation of TxScriptValidity for pending transactions

- Tx is used for both pending and in-ledger transactions.
- Having a TxScriptValidity entry for pending transactions doesn't make sense.
  - So change type to "Maybe TxScriptValidity", which we map to "Nothing" for
    pending transactions.
  - Further reduce TxScriptValidity to just "valid/invalid", mapping
    "unsupported" to "Nothing" too.
  - We think mapping "pending/unsupported" to "Nothing" is ok because the only
    case the wallet needs to distinguish is transactions that have explicitly
    failed validation.

sevanspowell committed:
Avoid use of variable-length indentation

sevanspowell committed:
Avoid multi-line parentheses

sevanspowell committed:
Better PersistFieldSql instance

sevanspowell committed:
Formatting

sevanspowell committed:
Use LambdaCase

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1952,8 +1958,13 @@ txHistoryFromEntity ti tip metas ins cins outs ws =
+               (txMetaFee m)
+               (txMetadata m)
+               (mkTxDerived m)
+               (txMetaScriptValidity m)
```
```suggestion
    mkItem m = mkTxWith
        (txMetaTxId m)
        (txMetaFee m)
        (txMetadata m)
        (mkTxDerived m)
        (txMetaScriptValidity m)
```

sevanspowell replied:
3 spaces?

jonathanknowles replied:
Apologies! I've edited my comment. 👍🏻

sevanspowell replied:
047085a247c0a21339dd41a6fc98a38dfe1a8b82

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1934,6 +1935,11 @@ mkTxMetaEntity wid txid mfee meta derived = TxMeta
+            case scriptValidity of
+                Nothing -> Nothing
+                (Just W.TxScriptValid) -> Just True
+                (Just W.TxScriptInvalid) -> Just False
```
```suggestion
    , txMetaScriptValidity = scriptValidity <&> \case
          W.TxScriptValid -> True
          W.TxScriptInvalid -> False
```

sevanspowell replied:
047085a247c0a21339dd41a6fc98a38dfe1a8b82

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1979,6 +1990,11 @@ txHistoryFromEntity ti tip metas ins cins outs ws =
+                case isValid of
+                    Nothing -> Nothing
+                    Just False -> Just W.TxScriptInvalid
+                    Just True -> Just W.TxScriptValid
```
```suggestion
            , W.txInfoScriptValidity = isValid <&> \case
                True -> W.TxScriptValid
                False -> W.TxScriptInvalid
```

sevanspowell replied:
047085a247c0a21339dd41a6fc98a38dfe1a8b82

jonathanknowles commented on lib/core/test/shared/Cardano/Wallet/DummyTarget/Primitive/Types.hs:
```diff
@@ -130,8 +136,9 @@ mkTx
+    -> Maybe TxScriptValidity
     -> Tx
-mkTx fees ins cins outs wdrls md =
+mkTx fees ins cins outs wdrls md isValid =
```
```suggestion
mkTx fees ins cins outs wdrls md validity =
```

sevanspowell replied:
4199c7f0925a78c92558361cad85135ee9bb85ea

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -2187,6 +2192,7 @@ instance Arbitrary (ApiTransaction n) where
             <*> arbitrary
             <*> pure txStatus
             <*> arbitrary
+            <*> arbitrary
```
Especially with these long record types, I think it's easier to figure out which generator is being used if we call it directly. (Otherwise, it requires the reader to work out which `Arbitrary` instance is being used.)
```suggestion
            <*> genTxScriptValidity
```

sevanspowell replied:
1105933ed6bcdd7d26ca9fb5d4ea4e7e7b143459

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:
```diff
@@ -376,10 +379,10 @@ arbitraryChainLength = 10
+    shrink (Tx _tid fees ins cins outs wdrls md isValid) =
+        [ mkTx fees ins' cins' outs' wdrls' md' isValid'
+        | (ins', cins', outs', wdrls', md', isValid') <-
+            shrink (ins, cins, outs, wdrls, md, isValid)
```
Suggestion: use a name that's consistent with the name of the type:

```suggestion
    shrink (Tx _tid fees ins cins outs wdrls md validity) =
        [ mkTx fees ins' cins' outs' wdrls' md' validity'
        | (ins', cins', outs', wdrls', md', validity') <-
            shrink (ins, cins, outs, wdrls, md, validity)
```

sevanspowell replied:
4199c7f

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:
```diff
@@ -388,7 +391,7 @@ instance Arbitrary Tx where
         wdrls <- fmap (Map.fromList . L.take 5) arbitrary
         fees <- arbitrary
-        mkTx fees ins cins outs wdrls <$> arbitrary
+        mkTx fees ins cins outs wdrls <$> arbitrary <*> arbitrary
```
I tend to think with these long record types, the calls to `arbitrary` make it harder than necessary to figure out exactly which generator is being used.

If we use the name of the generator directly, then it's easy for a reader to navigate to the source:

Suggestion:
```suggestion
        mkTx fees ins cins outs wdrls <$> arbitrary <*> genTxScriptValidity
```

sevanspowell replied:
1105933ed6bcdd7d26ca9fb5d4ea4e7e7b143459

sevanspowell committed:
Clean up pattern matches

sevanspowell committed:
Formatting

sevanspowell committed:
Use explicit TxScriptValidity generators

sevanspowell committed:
isValid -> validity

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -496,6 +503,8 @@ prop_changeUTxO_inner pendingTxs =
         , UTxO.size utxoAll == F.sum (F.length . view #outputs <$> pendingTxs)
         ]
+    & counterexample ("size utxoAll = " <> show (UTxO.size utxoAll))
+    & counterexample ("sum outputs = " <> show (F.sum (F.length . view #outputs <$> pendingTxs)))
```
https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#limit-line-length-to-80-characters

Suggestion: use the `report` combinator, as this handles pretty-printing for you (no need to call `show`).

```suggestion
    & report                                                                                                                                                                        
        (UTxO.size utxoAll)                                                                                                                                                         
        "UTxO.size utxoAll"                                                                                                                                                         
    & report                                                                                                                                                                        
        (F.sum (F.length . view #outputs <$> pendingTxs))                                                                                                                           
        "F.sum (F.length . view #outputs <$> pendingTxs)"    
```

jonathanknowles replied:
PR https://github.com/input-output-hk/cardano-wallet/pull/2876 moves `report` to a common location, making it easy to reuse here.

sevanspowell replied:
1105933ed6bcdd7d26ca9fb5d4ea4e7e7b143459

sevanspowell committed:
Fix HLint

sevanspowell committed:
Use Test.QuickCheck.Extra.report

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -3637,3 +3639,11 @@ instance FromJSON (ApiT (Script KeyHash)) where
+
+instance FromJSON (ApiT TxScriptValidity) where
+    parseJSON = fmap ApiT . genericParseJSON (Aeson.defaultOptions
+        { constructorTagModifier = camelTo2 '_' . drop 8 })
```
The parentheses can be removed (as this is record update syntax):
```suggestion
    parseJSON = fmap ApiT . genericParseJSON Aeson.defaultOptions
        { constructorTagModifier = camelTo2 '_' . drop 8 }
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -3637,3 +3639,11 @@ instance FromJSON (ApiT (Script KeyHash)) where
+
+instance ToJSON (ApiT TxScriptValidity) where
+    toJSON = genericToJSON (Aeson.defaultOptions
+        { constructorTagModifier = camelTo2 '_' . drop 8 }) . getApiT
```
Likewise:

```suggestion
    toJSON = genericToJSON Aeson.defaultOptions
        { constructorTagModifier = camelTo2 '_' . drop 8 } . getApiT
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -798,6 +803,27 @@ instance PersistField DerivationPrefix where
+        PersistBool True -> Right TxScriptValid
+        PersistBool False -> Right TxScriptInvalid
+        x -> Left $ T.unwords
+            [ "Failed to parse Haskell type `ScriptValidation`;"
```
```suggestion
            [ "Failed to parse Haskell type `TxScriptValidity`;"
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/Sqlite/TypesSpec.hs:
```diff
@@ -72,6 +74,7 @@ spec = do
         persistRoundtrip $ Proxy @StdGen
         persistRoundtrip $ Proxy @(Nested TxMetadata)
         persistRoundtrip $ Proxy @(Simple TxMetadata)
+        persistRoundtrip $ Proxy @TxScriptValidity
```
👍🏻

jonathanknowles committed:
Fix some formatting issues.

In this commit we:

- Remove variable indentation, occasionally using named variables to help.
- Wrap expressions that are significantly longer than the line limit.
- Use standard formatting for `where` clauses.
- Remove extraneous blank lines.

jonathanknowles committed:
Correct typo in comment for `liftShrink7`.

jonathanknowles committed:
Extract out function `utxoFromUnvalidatedTx` from `utxoFromTx`.

Use this function to simplify `changeUTxO`.

sevanspowell committed:
Small fixups

jonathanknowles committed:
Use `Arbitrary` instances to simplify unit tests.

Here we don't need to manually use `forAllShrink`, because we're just
using the standard generators and shrinkers without modification. We can
save a bit of boilerplate by using `Arbitrary` instances.

jonathanknowles committed:
Add coverage checks to properties in `Primitive.ModelSpec`.

jonathanknowles committed:
Make `failedScriptValidation` take a `Tx` argument.

The `failedScriptValidation` function is currently always called in a
context where the transaction itself is available.

Therefore, we can save callers the trouble of manually lensing into the
`scriptValidity` field, and move this lensing into the
`failedScriptValidation` function itself.

jonathanknowles committed:
Merge pull request #2879 from input-output-hk/jonathanknowles/adp-1036/collateral-chain-following

Additions to PR 2856

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -459,10 +478,34 @@ data TransactionInfo = TransactionInfo
+    -- ^ Tag indicating whether non-native scripts in this transaction are
+    -- expected to validate. This is added by the block creator when
+    -- constructing the block. May be Nothing for pre-Alonzo and pending
+    -- transactions.
```
```suggestion
    -- ^ Tag indicating whether non-native scripts in this transaction passed
    -- validation. This is added by the block creator when constructing the
    -- block. May be 'Nothing' for pre-Alonzo and pending transactions.
```

sevanspowell replied:
388c812

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -522,8 +524,101 @@ fileModeSpec =  do
+                        -- We got 12 from the faucet
+                        , (Incoming, 12)
+                        ]
+                    getAvailableBalance db `shouldReturn` 4
```
👍🏻

jonathanknowles committed:
Add roundtrip test for `ApiT TxScriptValidity`.

sevanspowell committed:
Update scriptValidity field comments

jonathanknowles committed:
Merge pull request #2880 from input-output-hk/jonathanknowles/roundtrip-test-api-tx-validity

Add roundtrip JSON encoding test for `ApiT TxScriptValidity`

jonathanknowles reviewed (approved):
LGTM! 👍🏻 

@sevanspowell thanks again for all your hard work on this!
